enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/embed.v1.ts

```diff
@@ -0,0 +1,170 @@
+  SdkIframeEmbedTagMessage,
+  SdkIframeEmbedTagSettings,
+} from "./types/embed";
+
+const EMBEDDING_ROUTE = "embed/sdk/v1";
+
+type EmbedSettingKey = keyof SdkIframeEmbedSettings;
+
+const ALLOWED_EMBED_SETTING_KEYS = [
+  "apiKey",
+  "instanceUrl",
+  "dashboardId",
+  "questionId",
+  "template",
+  "theme",
+  "locale",
+] as const satisfies EmbedSettingKey[];
+
+type AllowedEmbedSettingKey = (typeof ALLOWED_EMBED_SETTING_KEYS)[number];
+
+class MetabaseEmbed {
+  static readonly VERSION = "1.0.0";
+
+  private _settings: SdkIframeEmbedTagSettings;
+  private _isEmbedReady: boolean = false;
+  private iframe: HTMLIFrameElement | null = null;
+
+  constructor(settings: SdkIframeEmbedTagSettings) {
+    if (!settings.apiKey || !settings.instanceUrl) {
+      raiseError("api key and instance url must be provided");
+    }
+
+    this._validateEmbedSettings(settings);
+    this._settings = settings;
+    this._settings._isLocalhost = this.getIsLocalhost();
+
+    this._handleMessage = this._handleMessage.bind(this);
+    this._setup();
+  }
+
+  /**
+   * Merge these settings with the current settings.
+   */
+  public updateSettings(settings: Partial<SdkIframeEmbedSettings>) {
+    if (!this._isEmbedReady) {
+      warn("embed settings must be ready before updating the settings");
+      return;
+    }
+
+    if (settings.instanceUrl) {
+      raiseError("instanceUrl cannot be updated after the embed is created");
+    }
+
+    this._validateEmbedSettings(settings);
+    this._setEmbedSettings(settings);
+  }
+
+  public destroy() {
+    if (this.iframe) {
+      window.removeEventListener("message", this._handleMessage);
+      this.iframe.remove();
+      this._isEmbedReady = false;
+      this.iframe = null;
+    }
+  }
+
+  private _setEmbedSettings(settings: Partial<SdkIframeEmbedSettings>) {
+    const allowedSettings = Object.fromEntries(
+      Object.entries(settings).filter(([key]) =>
+        ALLOWED_EMBED_SETTING_KEYS.includes(key as AllowedEmbedSettingKey),
+      ),
+    );
+
+    this._settings = { ...this._settings, ...allowedSettings };
+    this._sendMessage("metabase.embed.setSettings", this._settings);
+  }
+
+  private _setup() {
+    const { instanceUrl, target, iframeClassName } = this._settings;
+
+    this.iframe = document.createElement("iframe");
+    this.iframe.src = `${instanceUrl}/${EMBEDDING_ROUTE}`;
+    this.iframe.style.width = "100%";
+    this.iframe.style.height = "100%";
+    this.iframe.style.border = "none";
+
+    if (iframeClassName) {
+      this.iframe.classList.add(iframeClassName);
+    }
+
+    window.addEventListener("message", this._handleMessage);
+
+    let parentContainer: HTMLElement | null = null;
+
+    if (typeof target === "string") {
+      parentContainer = document.querySelector(target);
+    } else if (target instanceof HTMLElement) {
+      parentContainer = target;
+    }
+
+    if (!parentContainer) {
+      raiseError(`cannot find embed container "${target}"`);
```

Comment:
should/could this be a part of `_validateEmbedSettings`? It just seems to me we spread condition checks everywhere.

Reply:
Yep! I just realized we can just validate after we merge the settings, so we can bundle all the condition checks in a single method [1]. Let me update this.

[1] https://github.com/metabase/metabase/pull/56915/files#r2087879224